docs/swift/DesignDoc.md

```diff
@@ -224,6 +245,16 @@ drafted or sent proposals for review that would benefit our designs:
 *   [Drafted: Resettable Properties]
     (https://github.com/patters/swift-evolution/blob/master/proposals/0000-resettable-properties.md)
 
+Lastly, we note some name transformations that we would apply to fields:
+
+* Repeated fields will have `Array` appended to their names. This is because
+  there is a trend in some messages to name repeated fields with the singular
+  version of the name (`descriptor.proto` in particular does this), and
+  `foo.name.append(bar)` looks potentially like appending a string to another
+  string whereas `foo.nameArray.append(bar)` is much clearer.
+* If a field name would collide with one of the properties or methods in
+  `ProtoMessage`'s public interface, we append `Field` to its name.
```

Comment:
(Doesn't the Objective-C implementation just append `Array` without an underscore?)

While inserting underscores would certainly disambiguate most cases, I *really* want to avoid adding underscores if I can avoid it--the resulting code would stick out like a sore among other Swift code.

While it would make the implementation more complex, I'd prefer to explore a phased approach where we detect collisions and only insert underscores where necessary. For example, if we had the following proto:

```protobuf
repeated int32 foo = 1;
optional int64 foo_array = 2;
```

We would process the singular fields first and collect their transformed names. In this case, we'd transform `foo_array` -> `fooArray`. Then, we'd process the repeated fields. We would try to transform `foo` -> `fooArray`, see the collision, and fallback `foo_Array` instead. That way, common cases still produce code that is as close to Swift conventions as possible, and only the uncommon cases suffer.

One potential problem is that a field's generated name could change by adding new fields to a proto: if you had repeated `foo` and later added singular `foo_array`, the new field would take the `fooArray` name and the previous one would change from `fooArray` to `foo_Array`. Is that something we should worry about?

With fields, we see the entire set of fields in the message so this should work--types, on the other hand, could come from multiple files and multiple `protoc` passes, so we probably have to unconditionally use underscores there.
Regarding leading/multiple (and trailing) underscores for Swift, I'd recommend retaining them--that seems like a situation where the proto author had a specific reason in mind for punctuating the field that way. Single inner underscores would be removed with the following letter uppercased.